Frontend/ViteJS native modules.md

```diff
@@ -0,0 +1,40 @@
+---
+
+tags:  frontend, native-modules, vitejs, esm
+
+author:  Tran Hoang Nam
+
+---
+Before ES modules was supported in the browser, we have no native mechanism for authoring [[JavaScript modules]] in a modularized fashion. That is why the concept of **"bundling"** and tools like webpack, Rollup and Parcel exists to improved the development experience for frontend developers.
+
+Howerver, when our project started to expand more, the amount of modules might increase from a hundred to thousands of modules which lead to performance bottleneck for **JavaScript-based** toolings. 
+
+**Vite** take full advantage of the availability of native ES modules in the browser and the rise of JavaScript tools written in compile-to-native languages by introducing `pre-bundles dependancies` using `esbuild` which was written in Go which 10-100x faster than JavaScript-based bundlers.
+
+---
+### Server built architecture 
+
+#### Bundle based dev server
+Bundle based dev server like Webpack built your application by combinning all the source-code and modules into a JavaScript-based bundle, every thing is done on the server-side and when you change something, the entire application have to build from the start.
+
+![](https://vitejs.dev/assets/bundler.37740380.png)
+
+#### Native ESM based dev server
+Vite appoarch was instead of bundling all everything on the server, it only bundle modules when the browser require it through HTTP request.
```

Comment:
```suggestion
Vite's approach was instead of bundling all everything on the server, it only bundles modules when the browser requires them through HTTP request.
```